Allow to set a custom ssh key in the cucumber test suite module

mackdk opened the pull request:
## What does this PR change?

This PR adds a configuration variable for the `cucumber_testsuite` module to allow setting an ssh key different from the default `~/.ssh/id_rsa.pub`. This prevents the `terraform apply` to fail due a different ssh key name.

juliogonzalez commented on modules/cucumber_testsuite/variables.tf:
```diff
@@ -19,6 +19,11 @@ variable "avahi_reflector" {
 }
 
+variable "ssh_key_path" {
+  description = "path of pub ssh key you want to use to access VMs, see libvirt/README.md"
```
`libvirt/README.md` does not exist. Do you mean `README_ADVANCED.md`?

mackdk replied:
The correct path to reference is ```backend_modules/libvirt/README.md```. It's fixed with the latest commit.

juliogonzalez commented:
@mackdk did you run a test? Maybe need to change the default value as I instructed, then pass your new parameters to the several places at `modules/cucumber_testsuite/main.tf` such as https://github.com/uyuni-project/sumaform/blob/master/modules/cucumber_testsuite/main.tf#L55

mackdk commented on modules/cucumber_testsuite/variables.tf:
```diff
@@ -19,6 +19,11 @@ variable "avahi_reflector" {
 
+variable "ssh_key_path" {
+  description = "path of pub ssh key you want to use to access VMs, see libvirt/README.md"
+  default     = "~/.ssh/id_rsa.pub"
```
If I understood this correctly, this is a different key. `./salt/controller/id_rsa.pub` is the one used by the controller to connect to the other machines, meanwhile the key I´ve added is the one used to connect from the host machine to the vm to configure them.

juliogonzalez replied:
Seems to be right, so now we have:
```
backend_modules/libvirt/README.md:All machines come with user `root` with password `linux`. They are also accessible via your SSH public key (by default `~/.ssh/id_rsa.pub`) if you have one.
```
And you want to make it configurable, right?

If so, I think you need to reference `backend_modules/libvirt/README.md`, and also document this new parameter there.

mackdk replied:
To clarify, this is the error I get while executing `terraform apply` and that this commit tries to address:

```
module.cucumber_testsuite.module.suse-minion.module.minion.module.host.null_resource.provisioning[0]: Creating...
module.cucumber_testsuite.module.suse-minion.module.minion.module.host.null_resource.provisioning[0]: Provisioning with 'file'...
module.cucumber_testsuite.module.suse-client.module.client.module.host.null_resource.provisioning[0]: Provisioning with 'remote-exec'...
module.cucumber_testsuite.module.suse-client.module.client.module.host.null_resource.provisioning[0] (remote-exec): Connecting to remote host via SSH...
module.cucumber_testsuite.module.suse-client.module.client.module.host.null_resource.provisioning[0] (remote-exec):   Host: 192.168.20.20
module.cucumber_testsuite.module.suse-client.module.client.module.host.null_resource.provisioning[0] (remote-exec):   User: root
module.cucumber_testsuite.module.suse-client.module.client.module.host.null_resource.provisioning[0] (remote-exec):   Password: true
module.cucumber_testsuite.module.suse-client.module.client.module.host.null_resource.provisioning[0] (remote-exec):   Private key: false
module.cucumber_testsuite.module.suse-client.module.client.module.host.null_resource.provisioning[0] (remote-exec):   Certificate: false
module.cucumber_testsuite.module.suse-client.module.client.module.host.null_resource.provisioning[0] (remote-exec):   SSH Agent: true
module.cucumber_testsuite.module.suse-client.module.client.module.host.null_resource.provisioning[0] (remote-exec):   Checking Host Key: false
module.cucumber_testsuite.module.suse-client.module.client.module.host.null_resource.provisioning[0] (remote-exec): Connected!
module.cucumber_testsuite.module.suse-minion.module.minion.module.host.null_resource.provisioning[0]: Provisioning with 'remote-exec'...
module.cucumber_testsuite.module.suse-minion.module.minion.module.host.null_resource.provisioning[0] (remote-exec): Connecting to remote host via SSH...
module.cucumber_testsuite.module.suse-minion.module.minion.module.host.null_resource.provisioning[0] (remote-exec):   Host: 192.168.20.46
module.cucumber_testsuite.module.suse-minion.module.minion.module.host.null_resource.provisioning[0] (remote-exec):   User: root
module.cucumber_testsuite.module.suse-minion.module.minion.module.host.null_resource.provisioning[0] (remote-exec):   Password: true
module.cucumber_testsuite.module.suse-minion.module.minion.module.host.null_resource.provisioning[0] (remote-exec):   Private key: false
module.cucumber_testsuite.module.suse-minion.module.minion.module.host.null_resource.provisioning[0] (remote-exec):   Certificate: false
module.cucumber_testsuite.module.suse-minion.module.minion.module.host.null_resource.provisioning[0] (remote-exec):   SSH Agent: true
module.cucumber_testsuite.module.suse-minion.module.minion.module.host.null_resource.provisioning[0] (remote-exec):   Checking Host Key: false
module.cucumber_testsuite.module.suse-minion.module.minion.module.host.null_resource.provisioning[0] (remote-exec): Connected!

Error: Error in function call: Call to function "file" failed: no file exists at /home/mackdk/.ssh/id_rsa.pub.

```

juliogonzalez replied:
Is this maybe https://github.com/uyuni-project/sumaform/blob/master/TROUBLESHOOTING.md#q-i-get-the-error--file-open-homesshid_rsapub-no-such-file-or-directory-in ?

I am not sure if this commit will achieve what you want or if a bigger rewrite will be needed to maintain the things consistent. You need @rjmateus here. I just added him as reviewer.

rjmateus replied:
For the first connection to start the configuration, sumaform doesn't use this certificate. It will connect to the IP address, user `root` and password `linux`.
After the set-up you should able to connect using the key you pass as parameter.
In this case, do you have any key, in your machine with the path `/home/mackdk/.ssh/id_rsa.pub`?
Sumaform is trying to read that file and load the certificate to the machine.
The default value looks good to me, since is the same we have on the base module.

mackdk replied:
> It will connect to the IP address, user `root` and password `linux`.

Oh, so it does not use the key to login but just to set up password-less login for the host, right?

> In this case, do you have any key, in your machine with the path `/home/mackdk/.ssh/id_rsa.pub`?

No, I don't. My ssh key uses the ED25519 algorithm. I introduced the parameter to be able to configure it from the `main.tf` and write:

```
module "cucumber_testsuite" {
  source = "./sumaform/modules/cucumber_testsuite"

  ssh_key_path = "~/.ssh/id_ed25519.pub"

  ...
```

Which is the same configuration that is supported for the main module.

rjmateus replied:
Did you get the error above with the configuration you just pasted? The error was pointing to the file `/home/mackdk/.ssh/id_rsa.pub`

mackdk replied:
No, sorry for the confusion. The error above is without the code from this PR, thus without setting the `ssh_key_path` in `main.tf` (because without this PR it is not possible to set it for the `cucumber_testsuite`).

By applying the changes in this PR and setting the correct path like I showed in my last comment, `terraform apply` works as expected.

rjmateus replied:
ok, now that makes sense :)

rjmateus commented:
> @mackdk did you run a test? Maybe need to change the default value as I instructed, then pass your new parameters to the several places at `modules/cucumber_testsuite/main.tf` such as https://github.com/uyuni-project/sumaform/blob/master/modules/cucumber_testsuite/main.tf#L55

I don't think it is needed. Deploy will merge both keys, and both will be available since the one introduced in here is being assigned to the base module.
Key from the base module is merged with key defined in the host module.

mackdk commented:
@juliogonzalez I updated the documentation to include a reference to the `ssh_key_path` variable and fixed the path pointing to the documentation in the variable descriptions.

juliogonzalez commented:
@mackdk since I don't have time to review now, but @rjmateus is happy with the changes then so am I as well.

Go ahead and merge :-)
